File: eval/go/simplekv.go
package main

import (
	"flag"
	"os"
	"fmt"
	"strconv"
	"hash/fnv"

	"github.com/vmware/go-pmem-transaction/pmem"
	"github.com/vmware/go-pmem-transaction/transaction"
)

const N int = 10

type pair struct {
	key   [32]byte
	idx   int
}

type data struct {
	buckets [][]pair
	values  []int
	magic   int
}

const (
	// A magic number used to identify if the root object initialization
	// completed successfully.
	magic = 0x1B2E8BFF7BFBD154
)

func hash(s string) int {
	h := fnv.New32a()
	h.Write([]byte(s))
	return int(h.Sum32())
}

func initialize(ptr *data) {
	txn("undo") {
		ptr.buckets = pmake([][]pair, N)
		ptr.magic = magic
	}
}

func get(ptr *data, key string) *int {
	index := hash(key) % N
	var bytes [32]byte
	copy(bytes[:], key)

	for i:=0; i<len(ptr.buckets[index]); i++ {
		e := ptr.buckets[index][i]
		if e.key == bytes {
			return &ptr.values[e.idx]
		}
	}

	return nil
}

func put(ptr *data, key string, val int) {
	index := hash(key) % N
	var bytes [32]byte
	copy(bytes[:], key)

	txn("undo") {
		/* search for element with specified key - if found
		 * transactionally update its value */
		for i:=0; i<len(ptr.buckets[index]); i++ {
			e := ptr.buckets[index][i];
			if e.key == bytes {
				ptr.values[e.idx] = val
				return
			}
		}

		/* if there is no element with specified key, insert new value
		 * to the end of values vector and put reference in proper
		 * bucket transactionally */
		l1 := len(ptr.values)
		if len(ptr.values) == 0 {
			ptr.values = pmake([]int, 0, 1)
		}
		ptr.values = append(ptr.values, val)

		if len(ptr.buckets[index]) == 0 {
			ptr.buckets[index] = pmake([]pair, 0, 1)
		}
		ptr.buckets[index] = append(ptr.buckets[index], pair {bytes, l1})
	}
}

func show_usage(prog string) {
	println("usage:", prog, "filename [get key|put key value]")

}

func main() {
	args := os.Args

	if len(args) < 4 {
		show_usage(args[0])
		return
	}

	var ptr *data
	flag.Parse()
	firstInit := pmem.Init(args[1])
	if firstInit {
		// first time run of the application
		ptr = (*data)(pmem.New("root", ptr))
		initialize(ptr)
	} else {
		// not a first time initialization
		ptr = (*data)(pmem.Get("root", ptr))

		// even though this is not a first time initialization, we should still
		// check if the named object exists and data initialization completed
		// succesfully. The magic element within the named object helps check
		// for successful data initialization.

		if ptr == nil {
			ptr = (*data)(pmem.New("root", ptr))
		}

		if ptr.magic != magic {
			initialize(ptr)
		}
	}

	if args[2] == "get" && len(args) == 4 {
		if n := get(ptr, args[3]); n != nil {
			fmt.Println(*n)
		} else {
			fmt.Println("No value found for", args[3])
		}
	} else if args[2] == "put" && len(args) == 5 {
		if n, err := strconv.Atoi(args[4]); err == nil {
			put(ptr, args[3], n)
		}
	} else if args[2] == "burst" && args[3] =="get" && len(args) == 5 {
		if m, err := strconv.Atoi(args[4]); err == nil {
			var v *int
			for i := 0; i < m; i++ {
				key := fmt.Sprintf("key%d", i);
				v = get(ptr, key)
			}
			if v != nil {
				fmt.Println("v =", *v)
			}
		}
    } else if args[2] == "burst" && args[3] == "put" && len(args) == 5 {
		if m, err := strconv.Atoi(args[4]); err == nil {
			for i := 0; i < m; i++ {
				key := fmt.Sprintf("key%d", i);
				put(ptr, key, i);
			}
		}
    } else {
        show_usage(args[0]);
    }
}

File: eval/mnemosyne/examples/simplekv/simplekv.cpp
#include <stdexcept>
#include <iostream>

#include "simplekv.hpp"
#include "pvar.hpp"

template <typename Value, std::size_t N>
const Value &simple_kv<Value, N>::get(const std::string &key) {
    auto index = std::hash<std::string>{}(key) % N;

    for (const auto &e: buckets[index]) {
        if (e.first == key) {
            return values[e.second];
        }
    }

    char msg[120];
    sprintf(msg, "no entry in simplekv for `%s`", key.c_str());
    throw std::out_of_range(msg);
}

template <typename Value, std::size_t N>
void simple_kv<Value, N>::put(const std::string &key, const Value &val) {
    auto index = std::hash<std::string>{}(key) % N;

    /* search for element with specified key - if found
        * transactionally update its value */
    for (const auto &e: buckets[index]) {
        if (e.first == key)
        {
            PTx { values[e.second] = val; }
            return;
        }
    }

    /* if there is no element with specified key, insert new value
        * to the end of values vector and put reference in proper
        * bucket transactionally */
    auto elem = std::pair<key_type, std::size_t>(key, values.size() - 1);
    PTx {
        values.push_back(val);        
        buckets[index].push_back(elem);
    }
}

void show_usage(char *argv[])
{
    std::cerr << "usage: " << argv[0]
              << " [get key|put key value] |"
              << " [burst get|put|putget count]" << std::endl;
}

void initialize() {
    int found = 0;
    PTx {
        if (PGET(kv) == NULL) {
            kv_t *k = (kv_t*)pmalloc(sizeof(kv_t));
            k->init();
            PSET(kv, k);
        } else {
            found = 1;
        }
    }
    if (!found) {
        fprintf(stderr, "Created the root object.\n");
    } else {
        fprintf(stderr, "Found the root object.\n");
    }
}

int main(int argc, char *argv[])
{
    if (argc < 2)
    {
        show_usage(argv);
        return 1;
    }

    initialize();

    if (std::string(argv[1]) == "get" && argc == 3) {
        kv_t *k = NULL;
        PTx { k = PGET(kv); }
        assert(k);
        std::cout << k->get(argv[2]) << std::endl;
    } else if (std::string(argv[1]) == "put" && argc == 4) {
        kv_t *k = NULL;
        PTx { k = PGET(kv); }
        assert(k);
        k->put(argv[2], std::stoi(argv[3]));
    } else if (std::string(argv[1]) == "burst" && std::string(argv[2]) == "get" && argc == 4) {
        kv_t *k = NULL;
        PTx { k = PGET(kv); }
        assert(k);
        int m = std::stoi(argv[3]);
        for (int i = 0; i < m; i++)
        {
            char key[32] = {0};
            sprintf(key, "key%d", i);
            k->get(key);
        }
    } else if (std::string(argv[1]) == "burst" && std::string(argv[2]) == "put" && argc == 4) {
        kv_t *k = NULL;
        PTx { k = PGET(kv); }
        assert(k);
        int m = std::stoi(argv[3]);
        fprintf(stderr, "inserting %d items...\n", m);
        for (int i = 0; i < m; i++)
        {
            char key[32] = {0};
            sprintf(key, "key%d", i);
            k->put(key, i);
        }
    } else {
        show_usage(argv);
    }

    return 0;
}

